### ISSUE TITLE
cannot backup anymore after upgrading from Ubuntu 20.10 to 21.04 (Invalid cross-device link)

### ISSUE THREAD

**Issue by Saroumane** [bit-team/backintime#1158]
After upgrade to Ubuntu 21.04 I get at each attempt to backup :

`ERROR: Failed to take snapshot !!!`

log says :

`Main profile Error: rsync: [generator] link "/media/root/BackupLocal/backintime/host/root/1/new_snapshot/backup/usr/share/javascript/jquery/jquery.js" => ../../20210504-070001-223/backup/usr/share/javascript/jquery/jquery.js failed: Invalid cross-device link (18)`

Indeed, the folder `jquery` is linked outside of the HDD backup :
```
# ll /media/root/BackupLocal/backintime/host/root/1/20210504-070001-223/backup/usr/share/javascript/
total 32
drwxr-xr-x   6 root root  4096 févr. 11 00:05 ./
drwxr-xr-x 359 root root 12288 avril  7 00:31 ../
lrwxrwxrwx  13 root root    21 juin  15  2020 font-awesome -> ../fonts-font-awesome/
lrwxrwxrwx  13 root root    29 nov.  15 00:14 jquery -> /usr/share/nodejs/jquery/dist/
drwxr-xr-x   9 root root  4096 nov.  15 00:02 mathjax/
drwxr-xr-x   2 root root  4096 avril 21  2019 prettify/
drwxr-xr-x   3 root root  4096 nov.  15 00:02 sphinxdoc/
drwxr-xr-x   2 root root  4096 avril 16 06:07 underscore/
```

I don't know what bug caused this offending link to appear, but I hope removing the link won't break everything :

`# rm /media/root/BackupLocal/backintime/host/root/1/20210504-070001-223/backup/usr/share/javascript/jquery`

**Comment by Hoimar** [comment]
Do you have any special settings configured in the Backintime configuration (like "copy unsafe links" or "copy symbolic links")?

And I think you can just remove (or move to another place outside the `backintime` folder) the offending `jquery` link, if anything, it will be copied again by the next backup job (when rsync detects that the link isnt't there anymore).

**Comment by Saroumane** [comment]
None of this.

**Comment by sputnik-1** [comment]
Hi @Saroumane 

Can you not just remove the offending cross-device link before you try to make a backup? I'm not sure if BIT provides a 'do not follow links' option?

Looking at Setting->Expert Options it looks like copying links is turned off by default. Has this been turned on by any chance?

![Screenshot_20210618_183307](https://user-images.githubusercontent.com/2807784/122597862-e65b1280-d063-11eb-9425-11069e779aba.png)

**Comment by Saroumane** [comment]
Hello @sputnik-1,  thanks for your concern.
Regarding your suggestion about the cross-link, I already did it (see https://github.com/bit-team/backintime/issues/1158#issue-880968900 ) and it worked, but does not explains how the bug was created.

Regarding your question about settings, @Hoimar asked here https://github.com/bit-team/backintime/issues/1158#issuecomment-852346950 and I answered there :  https://github.com/bit-team/backintime/issues/1158#issuecomment-852439396 that I have not modified these defaults settings.

**Comment by sputnik-1** [comment]
According to this Wikipedia article rsync is written in 'C'

https://en.wikipedia.org/wiki/Rsync

That's probably where this error is coming from.

All I can suggest is that you install the Eclipse CDT (C/C++ Development Tooling) from here:
https://www.eclipse.org/cdt/

and then step through the rsync code using the same command line that BIT uses to invoke rsync to see what's causing this error with the hard links?

**Comment by Saroumane** [comment]
For the record, I just reproduced this bug on another HDD which I backup every 6 months with a different profile in BiT (but same settings)
`Remote Error: rsync: [generator] link "/media/user/Backuphost/backintime/host/root/2/new_snapshot/backup/usr/share/javascript/jquery/jquery.js" => ../../20210126-085330-388/backup/usr/share/javascript/jquery/jquery.js failed: Invalid cross-device link (18)`
Except

**Comment by sputnik-1** [comment]
As a work-around, could you not add the full path to the directory containing the offending cross-link to your exclude tab settings - as a directory to ignore?

I did a live test backup of about 74GiB Tuesday night that took 2 hours to run, and only got a few errors as shown in the dual monitor screenshot.

Selecting /home/spock/.protonvpn-cli and BIT also kindly asked me if I wanted to add that to the ignore list, which it did as shown in the settings exclude tab.
HTH

![Screenshot_20210702_192626](https://user-images.githubusercontent.com/2807784/124315880-98b6cd80-db6c-11eb-9d76-f227613d1c38.png)

**Comment by Saroumane** [comment]
Sorry too late, I already applied my previous workaround : delete the offending (cross)link.
But that is very interesting : I understand these kind of errors should not cancel the whole backup process, I should end like you with a (partial) "backup with errors" ?
And after a backup with errors, the user should relaunch BiT, and for each folder triggering an error, select it the browsing tab (right part of BiT), and get a pop-up to ignore the folder ? (For the record, I never had "permission errors", as I run BiT as root)

**Comment by Hoimar** [comment]
> But that is very interesting : I understand these kind of errors should not cancel the whole backup process, I should end like you with a (partial) "backup with errors" ?

Chiming in here: You can enable "Continue on error" in the backup profile options.

> And after a backup with errors, the user should relaunch BiT, and for each folder triggering an error, select it the browsing tab (right part of BiT), and get a pop-up to ignore the folder ? (For the record, I never had "permission errors", as I run BiT as root)

Yes, you can exclude folders via right click in the browsing panel, but you can also pull up the snapshot log, select the offending folder path and rightclick on that to exclude.

**Comment by sputnik-1** [comment]
That's what I did - I selected the directory to exclude from inside the snapshot log, and then BIT offered to add it to the exclude list.

> "Chiming in here: You can enable "Continue on error" in the backup profile options."

That was enabled in my preferences, so although there were some errors with the first live backup, there were only a handful of those, and the backup copied all my important files.

The other things causing the errors were of no importance at all - so I fired up a root terminal and deleted the cause of those errors.

I ran another backup last night and it completed OK, in very good time as well.
The first initial backup took about 2.5 hours.

Before running BIT I quit Firefox web browser and also Thunderbird email client, as I did not want those programs changing any of their files during the backup process.

The second backup I ran last night completed without any of those errors.
I also noticed that the some Firefox and Thunderbird data files were copied again, due to being changed when those programs were running.
